File: python/Treap.py
# treap; binary search tree where each node has a random priority 
# maintained in a heap order based on these priorities 

import random

class Node:
  def __init__(self, key):
    self.key = key
    self.priority = random.randin(0, 1000)
    self.left = None
    self.right = None

class Treap:
  def __init__(self):
    self.root = None

  def insert(self, key):
    self.root = self._insert(self.root, key)
  def _insert(self, node, key):
    if node is None:
      return Node(key)
    if key < node.key:
      node.left = self._insert(node.left, key)
      if node.left.priority > node.priority:
        node = self.right_rotate(node)
    else:
      node.right = self._insert(node.right, key)
      if node.right.priority > node.priority:
        node = self.left_rotate(node)
    return node

  def right_rotate(self, x):
    y = x.left
    x.left = y.right
    y.right = x
    return y

  def left_rotate(self, x):
    y = x.right
    x.right = y.left
    y.left = x
    return y 

File: java/Treap.java
import java.util.Random;

public class Treap<T extends comparable<T>>
  {
    private Node root;
    private Random random = new Random();

    private class Node
      {
        T key;
        int priority;
        Node left, right;

        public Node(T key) {
          this.key = key;
          this.priority = random.nextInt();
        }
      }
    public void insert(T key)
    {
      root = insert(root, key);
    }
    private Node insert(Node node, T key)
    {
      if(node == null)
      {
        return new Node(key);
      }
      if(key.compareTo(node.key) < 0) {
        node.left = insert(node.left, key);
        if(node.left.priority > node.priority) {
          node = rightRotate(node);
        }
        else
        {
          node.right = insert(node.right, key);
          if(node.right.priority > node.priority)
          {
            node = leftRotate(node);
          }
        }
        return node;
      }
    }
      private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    private Node leftRotate(Node x)
    {
      Node y = x.right;
      x.right = y.left;
      y.left = x;
      return y;
    }
  }
